Experimental support for font hinting

AaronVanGeffen opened the pull request:
Traditionally, RCT2 has used a sprite-based font to render its strings. Since OpenRCT2's inception, support has been added for rendering with *TrueType* fonts to accomodate languages that do not use a Western script, such as Chinese, Korean, and Japanese.

In the past, OpenRCT2's palette has been a point of frustration, as it prevented implementation of direct font hinting ('anti-aliasing'), leading the Korean community to raise their minimum font size from 8 to 11 points (cf. #2317). This is not an ideal solution, however, as it leads text strings to overflow from their containing UI elements, since widgets are currently not flexible.

This PR, while a work in progress, is a crude attempt of implementing support for font hinting in OpenRCT2. Let me start off with some examples.

## Screenshots

(Before on the left, after on the right.)

![scenario_ru_before](https://user-images.githubusercontent.com/604665/30746257-7ee2ce34-9fa9-11e7-91f7-ef79ef5ca252.png) ![scenario_ru_after](https://user-images.githubusercontent.com/604665/30746254-7edf76b2-9fa9-11e7-868e-9c70df2631e2.png)

![settings_jp_before](https://user-images.githubusercontent.com/604665/30746255-7ee03de0-9fa9-11e7-8c06-db2c9f5e9ad0.png) ![settings_jp_after](https://user-images.githubusercontent.com/604665/30746256-7ee03afc-9fa9-11e7-93d5-a466711530e9.png)

![settings_ru_before](https://user-images.githubusercontent.com/604665/30746259-7ee673fe-9fa9-11e7-83a9-e28ba360dfb7.png) ![settings_ru_after](https://user-images.githubusercontent.com/604665/30746258-7ee50b68-9fa9-11e7-884e-af06436bdea3.png)

## Known issues

The current implementation does not play nicely with all window background colours — in some cases even overflowing into the wrong colour family entirely. I am hopeful we can resolve this as my understanding of the palette system deepens, however.

I have been eyeing `ColourMapA` in colour.c — it seems to be exactly what is needed. One caveat: in some cases a darker shade is needed, while the others require a lighter one. Outlined texts, like in window titles, are probably related here.

## In conclusion

Ideas or other contributions would be most appreciated! I'm stuck right now, but excited for the possibilities, so I thought I'd put it the code out here for now.

janisozaur commented:
We already have `[font]` section in `config.ini` describing various font properties, can the shaded v. solid option be included there as well?

Gymnasiast commented:
And if you add it, probably best to turn it on by default (assuming it won't be used for the sprite font).

marijnvdwerf commented:
Maybe we could create a function (with a cache) that calculates the blend between two colours. And then gets the closest palette index based on the closest colour.

AaronVanGeffen commented:
@janisozaur Making it optional was fairly straightforward (da4dae6cbd3fec19d8ee8c42e66d9ea95542b467), although I'm not too keen on the added compile-time dependency. Perhaps it's no big deal, though?

@marijnvdwerf Hmm, I was hoping to avoid blending altogether, but rather work from the background colour. Depending on the intensity of the foreground colour, we'd either use a darker or lighter shade. Would clamping on the colour palette in such a fashion be feasible?

AaronVanGeffen commented:
Getting there. Just to give an update, things look like this as of 6130d74e3972c276ea903567ad984430fad1828e:

![screen shot 2017-09-24 at 01 33 43](https://user-images.githubusercontent.com/604665/30777986-920265fe-a0c8-11e7-8b56-32dd6aabe1f3.png)

AaronVanGeffen commented:
Getting there. Here's what things look like as of e3091079ed5c3145ee8be26da9434bef31319142:

![screen shot 2017-09-24 at 01 44 00](https://user-images.githubusercontent.com/604665/30778029-2dcf712e-a0ca-11e7-82df-10d385159f28.png)

janisozaur commented:
@telk5093 @YJSoft @daihakken @izhangfei can you guys check how it works for you?

YJSoft commented:
![2017-09-25 18-14-22](https://user-images.githubusercontent.com/5893514/30801175-5ffb5ede-a21d-11e7-9cf8-9f4b59ff119f.png)
![image](https://user-images.githubusercontent.com/5893514/30800956-a709471a-a21c-11e7-8b92-57cf6e8a9c0d.png)

Looks OK except IME part(where background is black)

tested with below config
```
[font]
file_name = "NanumGothic.ttf"
font_name = "NanumGothic"
x_offset = 0
y_offset = -2
size_tiny = 9
size_small = 11
size_medium = 12
size_big = 13
height_tiny = 14
height_small = 14
height_medium = 14
height_big = 20
enable_hinting = true
```

AaronVanGeffen commented:
Thank you for testing, @YJSoft. I'm sure at one point the IME looked alright, so I should be able to fix it.

Gymnasiast commented:
@AaronVanGeffen Any update on fixing the IME?

AaronVanGeffen commented:
I had a quick look the other day, and while I'm sure I can fix it, it's a little less trivial than I assumed. Regrettably, I'm a little busy this week, but I'm hopeful can have another go at it this weekend.

AaronVanGeffen commented:
The IME issue should be fixed with 4bff727 adding support for black backgrounds. @YJSoft, could you confirm?

While doing some more thorough testing myself (sparking #6372, since merged), I encountered two more rare cases: the bright red text displaying losses in the finances window, and the light sea green colour in the theme window. Regrettably, I've had to add exceptions for both of them.

I think this is ready for review and testing, once more.

Gymnasiast commented on src/openrct2/drawing/string.c:
```diff
@@ -812,6 +812,12 @@ static void ttf_draw_string_raw_ttf(rct_drawpixelinfo *dpi, const utf8 *text, te
                             } else if (*dst < 10) {
                                 // For black backgrounds, hint using grey pixels.
                                 *dst = 13;
+                            } else if (colour == 153) {
```
Is there no constant for this? I remember @marijnvdwerf replacing all instances.

AaronVanGeffen replied:
`interface/colour.h` has a few colour constants, but not for any of these. Please correct me if I'm overlooking any, of course.

Gymnasiast replied:
I think you need to add the appropriate PALETTE_INDEX_x entries in the enum, but I'd like to hear from @marijnvdwerf first.

marijnvdwerf replied:
Either use `ColourMap`, or add a new `PALETTE_INDEX` entry.

AaronVanGeffen replied:
I've applied `ColourMapA` for now, where possible.  Suggestions for the other cases are welcome, too... I tried fiddling with the `BASE_COLOUR` macro as well, but I'm pressed for time and it wasn't quite working out yet, so I've left that out for now.

YJSoft commented:
<img width="266" alt="screenshot 2017-10-07 07 02 48" src="https://user-images.githubusercontent.com/5893514/31306830-a214cd90-ab92-11e7-8604-3f6110036762.png">

some letters are still hard to read, but I think it's font file's problem(letter too complex for small sizes)

AaronVanGeffen commented:
Thanks for testing! You're right that the glyphs are a little complicated for this font size, but there may be something we can do still — I have an idea. Could you check back here in, say, 15 minutes?

AaronVanGeffen commented:
@YJSoft Does abe79b7a937bed427ae285d33f0ffbe6fd006773 make things a little better for you? (It doesn't seem to make much of a difference when I test Japanese, but it's worth a shot...)

YJSoft commented:
<img width="262" alt="screenshot 2017-10-07 07 49 00" src="https://user-images.githubusercontent.com/5893514/31307143-ad9b9de6-ab98-11e7-8430-3bed7943530a.png">
much better

AaronVanGeffen commented:
Thanks again! It still doesn't look ideal, but at least the strokes are all visible, I guess.

@Gymnasiast, @marijnvdwerf Any more suggestions for improvement?

Gymnasiast commented on src/openrct2/drawing/scrolling_text.c:
```diff
@@ -1567,7 +1567,9 @@ void scrolling_text_set_bitmap_for_ttf(utf8 *text, sint32 scroll, uint8 *bitmap,
 
                 for (sint32 y = 0; y < height; y++) {
-                    if (src[y * pitch + x] != 0) *dst = colour;
+                    if (src[y * pitch + x] > 92 || (src[y * pitch + x] != 0 && !gConfigFonts.enable_hinting)) {
```
Brace placement

Gymnasiast commented on src/openrct2/drawing/ttf_sdlport.c:
```diff
@@ -1261,6 +1281,145 @@ TTFSurface *TTF_RenderUTF8_Solid(TTF_Font *font,
+    dst_check = (uint8*)textbuf->pixels + textbuf->pitch * textbuf->h;
+
+    /* Fill the palette with NUM_GRAYS levels of shading from bg to fg */
+    /*
```
Commented-out code; should be removed.

Gymnasiast commented on src/openrct2/drawing/ttf_sdlport.c:
```diff
@@ -1261,6 +1281,145 @@ TTFSurface *TTF_RenderUTF8_Solid(TTF_Font *font,
+    */
+
+    /* check kerning */
+    use_kerning = FT_HAS_KERNING( font->face ) && font->kerning;
```
`FT_HAS_KERNING(font->face)`

Gymnasiast commented:
The formatting needs a few kicks.

AaronVanGeffen commented:
The formatting concerns have been addressed per 86f28979c5b8bfd1c8d89288de37ba53cb70f4ff. I addressed each of the commits individually; three commits have been squashed into others to remove some experimental values. Effectively keeping the end results, while preserving the history of this PR over time.

I have not added a line to the changelog yet, as I expect merge conflicts there. Give me a sign if you think this is ready for merge, and I'll happily add it.

AaronVanGeffen commented:
Your formatting concerns have been addressed per 822286afc0816f986fec93eae20ebbe22be7b373. I have addressed each of the commits individually; three commits have been squashed into others to remove some experimental values. Effectively keeping the end results, while preserving the history of this PR over time.

I have not added a line to the changelog yet, as I expect merge conflicts there. Give me a sign if you think this is ready for merge, and I will happily add it.

marijnvdwerf commented on src/openrct2/drawing/scrolling_text.c:
```diff
@@ -1566,8 +1566,12 @@ void scrolling_text_set_bitmap_for_ttf(utf8 *text, sint32 scroll, uint8 *bitmap,
-                    if (src[y * pitch + x] != 0) *dst = colour;
+                for (sint32 y = 0; y < height; y++)
+                {
+                    if (src[y * pitch + x] > 92 || (src[y * pitch + x] != 0 && !gConfigFonts.enable_hinting))
```
What is `92`?

AaronVanGeffen replied:
Here, it's an arbitrary threshold for when to interpret a greyscale pixel as a monochrome one for the banners. As my changes change the rendering function to work on the domain [0, 255] rather than [0, 1], a threshold is needed to stick with a monochrome palette for the banners.

I think I fiddled around with it in steps of 8 or 12. It could probably be 90 or 95 as well; we're obviously not bound by RCT2's palette as far as the `src` bitmap is concerned...

marijnvdwerf commented on src/openrct2/drawing/string.c:
```diff
@@ -798,7 +799,44 @@ static void ttf_draw_string_raw_ttf(rct_drawpixelinfo *dpi, const utf8 *text, te
+                            {
+                                // For other texts, hinting direction depends on the intensity of the
+                                // foreground colour.
+                                bool intensive = colour % 12 > 4 && colour % 12 < 10;
```
Do these numbers mean anything, or are they the result of experimentation?

AaronVanGeffen replied:
`colour % 12` gets you the palette row in question, as you probably figured. I found the domain between 4 and 10 through experimenting — these seem to work reasonably well across the board, save for the few exceptions as implemented.

marijnvdwerf commented on src/openrct2/drawing/ttf_sdlport.c:
```diff
@@ -1261,6 +1282,146 @@ TTFSurface *TTF_RenderUTF8_Solid(TTF_Font *font,
+TTFSurface *TTF_RenderUTF8_Shaded(TTF_Font *font,
+                const char *text, uint32 fg, uint32 bg)
+{
+    bool first;
```
@IntelOrca Do we do variable predeclaration if the files come from an external source?

IntelOrca replied:
I don't know.

AaronVanGeffen replied:
Save for @Gymnasiast's comments on formatting, I stuck to the implementation of the counterparts that were already there — which did leave the predeclaration intact.

Personally, I think it might come in handy not to refactor the SDL2_ttf code too much, if we ever want to merge changes from upstream.

Gymnasiast replied:
I think predeclarations are fine in some cases. This is one.

AaronVanGeffen commented:
@marijnvdwerf Cheers for the in-depth review. You're right to be weary of arbitrary thresholds and the likes, of course. On Gitter, you mentioned you were working on proper colour blending to replace the approach I've taken in this PR. That's something I most welcome, of course. If it's close to being finished, I'll happily incorporate it into this PR as well, dropping my 'hackier' commits. If not, then perhaps you could include it in #6075 instead, for both our sanities.

marijnvdwerf commented on src/openrct2/drawing/string.c:
```diff
@@ -739,7 +739,7 @@ uint8 findClosestIndex(uint8 red, uint8 green, uint8 blue)
     return closest;
 }
 
-uint8 blend(const uint8 base, const uint8 layer)
```
Naming could probably be improved upon. As well as pointing out that the values are palette indexes.

AaronVanGeffen replied:
Agreed, it's probably not clear what the purpose of a 'blend' function in string.c. I was thinking it might be better to export these functions from colour.{c,h} instead, and move the blend table there as well. What are your thoughts on that?

AaronVanGeffen commented:
The new blending code introduced by @marijnvdwerf in 799e202bec621fba375c682d19697f948b6f46fa works quite well overall, and is generally an improvement, especially conceptually. However, it did not account for the remappable colours in the palette; I've introduced `isRemappableColour` in b4941b9a56ff47aa9a7d37869ee57c22a71f38ae to address this to counter flickering text.

There is one more regression I think should be addressed before merging. I'm not sure how to solve it yet, however. Consider the scenario window on 1c5623fe091984b795316328bd5ba2e770843dd5:

![screenshot_20171013_203900](https://user-images.githubusercontent.com/604665/31561209-6fbd939e-b057-11e7-8ba2-b939025ac961.png)

Here's the same window on b4941b9a56ff47aa9a7d37869ee57c22a71f38ae:

![screenshot_20171013_202659](https://user-images.githubusercontent.com/604665/31561208-6f9f47fe-b057-11e7-99a5-c13e361fc6ec.png)

As you can see, the glyphs are rendered much more thinly, making them harder to read.

marijnvdwerf commented:
![image](https://user-images.githubusercontent.com/737603/31563674-654630ca-b060-11e7-8ad7-c1a23bdb08fd.png)
The only dynamic palette is the stuff following and including 0xE6 (360)

AaronVanGeffen commented:
Ah, I see. I was going by the [palette spreadsheet](https://docs.google.com/spreadsheets/d/1OtTAytgtecqW7bbN8K6lb6dw4LYUMiXXBb7ZLGyk__U), but I considered all remapping areas rather than just the ones that are shifted dynamically. Will adjust the relevant commit.

marijnvdwerf commented:
Tofu renders pretty badly:
![image](https://user-images.githubusercontent.com/737603/31570366-4f10f880-b083-11e7-8b21-d8986dbc7b3b.png)

AaronVanGeffen commented:
Hmm, that tofu renders badly is unfortunate, but I wonder whether it's worth it overall. I sent this over Gitter earlier, but here's a screenshot comparing the 'mono' and 'light' hinting modes; 'normal' produces similar results to 'mono'.

![comparison](https://user-images.githubusercontent.com/604665/31573664-93de8724-b0c0-11e7-915b-c98f2b1ce1d5.png)

As you can see, especially the more complicated glyphs are rendered much more clearly, while they are still adequately hinted. Hence, for Japanese, I would strongly prefer 'light' over 'mono'.

AaronVanGeffen commented:
@marijnvdwerf I think I've fixed it. As it turns out, the tofu rendering was fine on Linux, but it did occur consistently on macOS. While I'd rather have avoided it, I've made the hinting threshold configurable in the font descriptor. That way, Arial's can be 40, while the CJK fonts enjoy 60.

Hence, tofu show up again as needed when Arial is used as a fallback font:

![tofu](https://user-images.githubusercontent.com/604665/31578757-02163fb8-b128-11e7-8771-4c7732c89bc8.png)

Personally, I'm very happy with the end result. Unless you have any more issues, I'll happily clean up the commit history so this can be merged.

marijnvdwerf commented on src/openrct2/drawing/ttf_sdlport.c:
```diff
@@ -1272,6 +1433,20 @@ static bool CharacterIsDelimiter(char c, const char *delimiters)
     return false;
 }
 
+void TTF_SetFontHinting(TTF_Font* font, int hinting)
```
This is only called with `0` or `1` for `hinting`.

AaronVanGeffen replied:
Yes, but it is the original function backported from SDL_ttf. I figured that might be better to use, conceptually, rather than just hacking it in.

marijnvdwerf commented on src/openrct2/interface/Fonts.cpp:
```diff
@@ -26,38 +26,38 @@
-    { "msgothic.ttc", "MS PGothic", 12, 1, 0, 17, nullptr },
-    { "msgothic.ttc", "MS PGothic", 12, 1, 0, 17, nullptr },
-    { "msgothic.ttc", "MS PGothic", 13, 1, 0, 20, nullptr },
+    { "msgothic.ttc", "MS PGothic", 9, 1, 0, 15, 60, nullptr },
```
Maybe add a define for these thresholds?

AaronVanGeffen replied:
Sure! Since this is C++, shall we implement them using `constexpr`? As for their names, what about `HINTING_THRESHOLD_LOW` for 40 and `HINTING_THRESHOLD_MEDIUM` for 60?

marijnvdwerf replied:
No idea what a `constexpr` is, but if it's better: sure :P. Names seem fine though.

AaronVanGeffen replied:
Actually, `constexpr` is probably overkill. I've just defined them as old-skool consts. :)

marijnvdwerf commented:
Seems fine to me. Probably should revert the logo though.

AaronVanGeffen commented:
Haha, yes, absolutely. Thanks. I'll do a rebase onto develop in the morning, squashing and dropping a commit here and there, then.

AaronVanGeffen commented on src/openrct2/interface/colour.c:
```diff
@@ -66,14 +66,14 @@ void colours_init_maps()
 {
     sint16 closest = -1;
-    sint32 closestDistance = INT32_MAX;
+    double closestDistance = INT32_MAX;
```
@marijnvdwerf Do you reckon we should drop the doubles in favour of going back to sint32s? They don't seem to have made much of an impact.

marijnvdwerf replied:
I've just ran a test on all colours, int or double always produce the same outcome.

AaronVanGeffen commented:
Rebased the lot on top of `develop`. Merged and dropped a few commits, too. Wondering whether I should squash 8c1ae5e651ebdb49a79e18e507e5da835aa60d86, 253b52573c00dd05c441b737c43b0df44b4cbb68, 899fe2faf967547416b669453a6248d080139961? Personally, I really like having the history around.

@marijnvdwerf, what do you think?

marijnvdwerf commented on src/openrct2/interface/colour.h:
```diff
@@ -148,6 +151,10 @@ extern rct_colour_map ColourMapA[COLOUR_COUNT];
 void colours_init_maps();
 
+#ifndef NO_TTF
+uint8 blend(const uint8 paletteIndex1, const uint8 paletteIndex2);
```
Can you rename this to something more verbose?

marijnvdwerf commented:
I'm fine with the history as-is. I'll merge it once you renamed the `blend` function 👍

AaronVanGeffen commented:
Done! Thanks for the help and review!
